server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -439,61 +402,332 @@ public static boolean tryFastFilterAggregation(
+        } catch (CollectionTerminatedException e) {
+            logger.debug("Early terminate since no more range to collect");
+        }
+        collector.finalizePreviousRange();
+
+        return debugInfo;
+    }
+
+    private static void intersectWithRanges(
+        PointValues.IntersectVisitor visitor,
+        PointValues.PointTree pointTree,
+        RangeCollectorForPointTree collector,
+        DebugInfo debug
+    ) throws IOException {
+        PointValues.Relation r = visitor.compare(pointTree.getMinPackedValue(), pointTree.getMaxPackedValue());
+
+        switch (r) {
+            case CELL_INSIDE_QUERY:
+                collector.countNode((int) pointTree.size());
+                debug.visitInner();
+                break;
+            case CELL_CROSSES_QUERY:
+                if (pointTree.moveToChild()) {
+                    do {
+                        intersectWithRanges(visitor, pointTree, collector, debug);
+                    } while (pointTree.moveToSibling());
+                    pointTree.moveToParent();
+                } else {
+                    pointTree.visitDocValues(visitor);
+                    debug.visitLeaf();
+                }
+                break;
+            case CELL_OUTSIDE_QUERY:
+        }
+    }
+
+    private static PointValues.IntersectVisitor getIntersectVisitor(
+        RangeCollectorForPointTree collector,
+        ArrayUtil.ByteArrayComparator comparator
+    ) {
+        return new PointValues.IntersectVisitor() {
+            @Override
+            public void visit(int docID) throws IOException {
+                // this branch should be unreachable
+                throw new UnsupportedOperationException(
+                    "This IntersectVisitor does not perform any actions on a " + "docID=" + docID + " node being visited"
+                );
+            }
+
+            @Override
+            public void visit(int docID, byte[] packedValue) throws IOException {
+                visitPoints(packedValue, collector::count);
+            }
+
+            @Override
+            public void visit(DocIdSetIterator iterator, byte[] packedValue) throws IOException {
+                visitPoints(packedValue, () -> {
+                    for (int doc = iterator.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = iterator.nextDoc()) {
+                        collector.count();
+                    }
+                });
+            }
+
+            private void visitPoints(byte[] packedValue, CheckedRunnable<IOException> collect) throws IOException {
+                if (comparator.compare(packedValue, 0, collector.activeRangeAsByteArray[1], 0) > 0) {
+                    // need to move to next range
+                    collector.finalizePreviousRange();
+                    if (collector.iterateRangeEnd(packedValue, this::compareByteValue)) {
+                        throw new CollectionTerminatedException();
+                    }
+                }
```

Comment:
This assumes non-overlapping ranges, right?

From what I can tell from the `createRangesFromAgg` method, you always guarantee that the buckets have the same width (`interval`), and the upper bound is basically always `low + interval`, then the next lower bound is `low + interval`, but you do some off-by-1 subtraction so they're not overlapping, right?

Reply:
Yes, the ranges are always non-overlapping. The -1 subtraction was done here:

```
// Subtract -1 if the minimum is roundedLow as roundedLow itself
// is included in the next bucket
long upper = i + 1 == bucketCount ? high : fieldType.convertRoundedMillisToNanos(roundedLow) - 1;
```

Reply:
> you always guarantee that the buckets have the same width

While this is the case right now, I don't think not having same width makes a difference for multi range traversal. Please correct me if needed since, we plan to use same optimization for range aggregation as well.

Reply:
Yes, that's the characteristics of current ranges. Long type (the type for date type) and connected with each other (previous upper bound +1 is next lower bound)
This characteristic may change when apply to range agg, so we can keep this thread open for reference later.